### ISSUE TITLE
Adopt a deprecation policy

### ISSUE THREAD

**Issue by felixfbecker** [semantic-release/semantic-release#680]
semantic-release has a lot of plugin hooks. As a plugin author, it's hard to follow the very frequent major releases of semantic-release. There is a risk of the "Major version cliff" as described by the React team: https://reactjs.org/blog/2016/02/19/new-versioning-scheme.html#avoiding-the-major-cliff

As users, we sometimes cannot update to the latest semantic-release because one of the plugins we depend on only works with the older semantic-release. As plugin authors, we want to support the semantic-release versions that users use. This can create a major version cliff where nobody updates, which hurts the ecosystem.

The solution to this is also outlined in the above blog post: making it easy for plugins to always support two major versions at the same time. That means always first deprecate a feature and add the alternative in a minor version, then remove the deprecated feature in a major. That way plugins can adopt the new feature and are ready for the next major while still supporting the previous version.

As a package that builds a lot on its plugin ecosystem it is good to explicitly follow such a policy.

Some examples of breaking changes in semantic-release and suggestions how they could have gone through such a policy:

#### 15.0.0
> Committing or creating files in the publish plugin hook is not supported anymore and now must be done in the prepare hook

Add the `prepare` plugin hook in 14.1.0 and deprecate creating files in the `publish` hook. `prepare` is called before `publish` but there are no other changes in behaviour. Make the breaking change in 15.0.0

#### 11.0.0

> Each plugin is expected to return an async function or a Promise returning function. The callback parameter is not passed to plugins anymore.

Allow plugins to return a Promise, but also pass a callback. Deprecate the callback and print a warning when it is called. Stop passing the callback in 12.0.0

#### 9.0.0

> Semantic-Release must now be executed with semantic-release instead of semantic-release pre && npm publish && semantic-release post.

When semantic-release is called with `semantic-release pre`, maintain the old behaviour (exit code) but print a deprecation warning. Remove `pre` and `post` in 10.0.0

**Comment by gr2m** [comment]
> always first deprecate a feature and add the alternative in a minor version, then remove the deprecated feature in a major

I agree that this is best practise. From my experience it doesn’t happen until there is a release process in place with different people being in charge for different parts. I'm not sure we can tackle it yet, but we sure will in future.

**Comment by pvdlg** [comment]
Maybe implementing #563 could help here. We could have the last major release on `@next` and the previous one on `@latest` and plugin could do the same. Just a though.

I tried to maintain backward compatibility as much as I could but it's not always easy and in some cases it could have been done in a better way.
But unfortunately in some cases it's not really possible or it creates a lot of extra work.

For example in the cases of `9.0.0` it's really difficult to maintain the ability to call semantic-release with both `semantic-release` and `semantic-release pre && npm publish && semantic-release post` because internally a lot has changed.

Maintaining backward compatibility (with deprecation) from one major to the other takes time. And this is time that would not be invested in developing new features, so it's a trade-off.

So far, developing new features was prioritized over maintaining backward compatibility, which create some hurdles. Hopefully the pace of breaking changes should slow down, or be less impactful. But as the number of users is growing it's probably time for us to move the cursor a bit more toward focus on backward compatibility.

**Comment by pvdlg** [comment]
We also try to release on `@latest` less often, and keep breaking changes on `@next` only for a longer time. This way we can release on `@latest` multiple breaking changes at once.
One easy thing we can start doing is to give advance notice when a version will be made available on `@latest`, so plugin creator have more visibility.

**Comment by felixfbecker** [comment]
> We also try to release on @latest less often, and keep breaking changes on @next only for a longer time. This way we can release on @latest multiple breaking changes at once.

Well I assume the goal of having an unstable (next) release channel is to have users try it out and report bugs before it gets promoted to stable (latest). But to have users try it out, plugins need to support it. That is hard when a plugin cannot be compatible with both versions at the same time. Plugins could of course also move to a next/latest release flow, but imo plugins shouldn't be required to have a release process as complicated as core.
